FIX make sure IterativeImputer does not skip iterative process when keep_empty_features=True

arifqodari opened the pull request:
#### Reference Issues

Fixes #29375 

**Empty Feature = Feature that consist exclusively of missing values.**

#### What does this implement/fix? Explain your changes.

This PR fixes 2 cases mentioned on this issue #29375:

1. In the `IterativeImputer`, when `keep_empty_features` is set to `True`, the iterative imputation step is skipped, and the result is constructed solely from the initial imputation (`SimpleImputer`). This occurs because `mask_missing_values` is set to `True` for all non-empty feature columns. As a result, the outcomes differ depending on whether `keep_empty_features` is set to `False` or `True`. The proposed solution is to remove this step and retain the original `mask_missing_values`.
2. When running the iterative imputation step with empty features, a `ValueError: Found array with 0 sample(s) ...` exception is raised because the estimator attempts to fit data with 0 rows. The proposed solution is to use the imputed values from the initial imputation result when the estimator encounters empty features. These values are constant across all rows.

#### Any other comments?

arifqodari committed:
Fix Iterative Imputation not Triggered when Keep Empty Features

arifqodari committed:
Merge branch 'main' into fix_iterative_imputer_keep_empty_features

glemaitre commented:
We will need some non-regression tests to check that we get the expected results. On the top of the head, we should test:

- A matrix without a column full of `np.nan` and check that `keep_empty_feature=True/False` lead to the same results. We can strengthen the current test in this regard.
- A matrix containing a entire `np.nan` column and check that other columns remain the same and check that get a zero column if `keep_empty_feature=True` otherwise, it should be drop.

So from the code, I think that the diff is not what I would expect. I would only expect a change in the initial imputation with something like:


```python
        if not self.keep_empty_features:
            # drop empty features
            valid_mask = np.flatnonzero(
                np.logical_not(np.isnan(self.initial_imputer_.statistics_))
            )
            Xt = X[:, valid_mask]
            mask_missing_values = mask_missing_values[:, valid_mask]
        else:
            # mark empty features as not missing and keep the original
            # imputation
            is_empty_feature = np.all(mask_missing_values, axis=0)
            mask_missing_values[:, is_empty_feature] = False
            Xt = X
            Xt[:, is_empty_feature] = X_filled[:, is_empty_feature]
```

Basically, `keep_empty_feature=False` remains the same behaviour as before. For the case of `keep_empty_feature`, we need to detect which column was full of `np.nan`. We can do that since we already computed the mask and we cannot use the `initial_imputer_` since it does show this information (it imputes the values with 0). Once we have the info, we can set the empty features as not missing and fill the `X` matrix with the filled value that is zero.

glemaitre commented:
In addition to the tests and the right fix, we need an entry in the changelog:

Please add an entry to the change log at `doc/whats_new/v1.5.rst`. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`.

arifqodari committed:
Added Two More Tests for IterativeImputer

arifqodari committed:
Fix Applied Only in the Initial Imputation Step

arifqodari committed:
Merge branch 'main' into fix_iterative_imputer_keep_empty_features

arifqodari committed:
Added Changelog

arifqodari committed:
Remove Trailing Whitespace and Unused Lines

arifqodari committed:
Added More Tests with Various Initial Strategis

arifqodari committed:
Fix Error in IterativeImputer When Initial Strategy Constant

arifqodari committed:
Fix Lint Issues

arifqodari commented:
> We will need some non-regression tests to check that we get the expected results. On the top of the head, we should test:
> 
>   * A matrix without a column full of `np.nan` and check that `keep_empty_feature=True/False` lead to the same results. We can strengthen the current test in this regard.
> 
>   * A matrix containing a entire `np.nan` column and check that other columns remain the same and check that get a zero column if `keep_empty_feature=True` otherwise, it should be drop.

Thanks for the feedback @glemaitre . I have added two more tests covering various `initial_strategy` as well.



> So from the code, I think that the diff is not what I would expect. I would only expect a change in the initial imputation with something like:
> 
> ```python
>         if not self.keep_empty_features:
>             # drop empty features
>             valid_mask = np.flatnonzero(
>                 np.logical_not(np.isnan(self.initial_imputer_.statistics_))
>             )
>             Xt = X[:, valid_mask]
>             mask_missing_values = mask_missing_values[:, valid_mask]
>         else:
>             # mark empty features as not missing and keep the original
>             # imputation
>             is_empty_feature = np.all(mask_missing_values, axis=0)
>             mask_missing_values[:, is_empty_feature] = False
>             Xt = X
>             Xt[:, is_empty_feature] = X_filled[:, is_empty_feature]
> ```
> 
> Basically, `keep_empty_feature=False` remains the same behaviour as before. For the case of `keep_empty_feature`, we need to detect which column was full of `np.nan`. We can do that since we already computed the mask and we cannot use the `initial_imputer_` since it does show this information (it imputes the values with 0). Once we have the info, we can set the empty features as not missing and fill the `X` matrix with the filled value that is zero.

I have updated the changes as suggested and apparently there is one case that is not covered. It is when the `initial_strategy="constant"`, the iterative process estimator attempts to fit data with all missing rows. So, I put another `if` condition if it is `in_fit` it will mark the empty features as not missing and use results from initial imputation. Let me know your thoughts @glemaitre .

Additionally, I have added one new entry to the 1.5.2 changelogs.

arifqodari committed:
Merge branch 'main' into fix_iterative_imputer_keep_empty_features

arifqodari committed:
Merge branch 'main' into fix_iterative_imputer_keep_empty_features

arifqodari committed:
Merge branch 'main' into fix_iterative_imputer_keep_empty_features

glemaitre commented on doc/whats_new/v1.5.rst:
```diff
@@ -53,6 +53,13 @@ Changelog
   platform-agnosticity of the estimator.
   :pr:`29612` by :user:`Olivier Grisel <ogrisel>`.
 
+:mod:`sklearn.impute`
```
Sorry, I was a bit optimistic that we merge before 1.5.2 that is going to be release this week.

Could you move this entry in 1.6.0 at the end.

arifqodari replied:
Done

arifqodari committed:
Move Changelog to 1.6

glemaitre commented on sklearn/impute/_iterative.py:
```diff
@@ -646,19 +646,27 @@ def _initial_imputation(self, X, in_fit=False):
-        )
-
         if not self.keep_empty_features:
+            valid_mask = np.flatnonzero(
+                np.logical_not(np.isnan(self.initial_imputer_.statistics_))
+            )
+
             # drop empty features
             Xt = X[:, valid_mask]
             mask_missing_values = mask_missing_values[:, valid_mask]
+
+            if in_fit:
+                is_empty_feature = np.all(mask_missing_values, axis=0)
+                mask_missing_values[:, is_empty_feature] = False
+                Xt[:, is_empty_feature] = X_filled[:, is_empty_feature]
+
         else:
             # mark empty features as not missing and keep the original
             # imputation
-            mask_missing_values[:, valid_mask] = True
+            is_empty_feature = np.all(mask_missing_values, axis=0)
+            mask_missing_values[:, is_empty_feature] = False
             Xt = X
+            Xt[:, is_empty_feature] = X_filled[:, is_empty_feature]
```
So I would expect the following:


```python

is_empty_feature = np.all(mask_missing_values, axis=0)

        if not self.keep_empty_features:
            # drop empty features
            Xt = X[:, ~is_empty_feature]
            mask_missing_values = mask_missing_values[:, ~is_empty_feature]

            if self.initial_imputer_.get_params()['strategy'] == 'constant':
                # The constant strategy has a specific behavior and preserve empty
                # features even with ``keep_empty_features=False``. We need to drop
                # the column for consistency.
                # TODO: remove this `if` branch once the following issue is addressed:
                # https://github.com/scikit-learn/scikit-learn/issues/29827
                X_filled = X_filled[:, ~is_empty_feature]
        else:
            # mark empty features as not missing and keep the original imputation
            mask_missing_values[:, is_empty_feature] = False
            Xt = X
            Xt[:, is_empty_feature] = X_filled[:, is_empty_feature]

```

arifqodari replied:
Tried with this change and got `IndexError: index 3 is out of bounds for axis 1 with size 3` on this test `test_iterative_imputer_skip_non_missing`.

Basically, it is when
- `X_train` has **no** empty features
- `X_test` has empty features
- `keep_empty_features=False`
- `skip_complete=True` or `False`

```python
X_train = np.array([[5, 2, 2, 1], [10, 1, 2, 7], [3, 1, 1, 1], [8, 4, 2, 2]])
X_test = np.array([[np.nan, 2, 4, 5], [np.nan, 4, 1, 2], [np.nan, 1, 10, 1]])
imputer = IterativeImputer(
    initial_strategy="mean", skip_complete=skip_complete, random_state=rng
)
X_test_est = imputer.fit(X_train).transform(X_test)
if skip_complete:
    # impute with the initial strategy: 'mean'
    assert_allclose(X_test_est[:, 0], np.mean(X_train[:, 0]))
else:
    assert_allclose(X_test_est[:, 0], [11, 7, 12], rtol=1e-4)
```

In `transform`, using only this 
```python
is_empty_feature = np.all(mask_missing_values, axis=0)
valid_mask = ~is_empty_feature
```
will remove the empty feature from `X_test` which I guess we don't expect since the `X_train` has no empty feature.

What do you think?

I can fix with these following changes:
```python
if in_fit:
    is_non_empty_feature = np.logical_not(is_empty_feature)
else:
    is_non_empty_feature = np.logical_not(np.isnan(self.initial_imputer_.statistics_))
``` 

But it still require `in_fit`. Let me know if you have another suggestion.

glemaitre replied:
> is_non_empty_feature = np.logical_not(np.isnan(self.initial_imputer_.statistics_))

I don't like so much this line because the `SimpleImputer` is not consistent. Maybe the cleaner way is to store the empty feature at `fit` and not recompute it otherwise:

```python
        if in_fit:
            self._is_empty_feature = np.all(mask_missing_values, axis=0)

        if not self.keep_empty_features:
            # drop empty features
            Xt = X[:, ~self._is_empty_feature]
            mask_missing_values = mask_missing_values[:, ~self._is_empty_feature]

            if self.initial_imputer_.get_params()['strategy'] == 'constant':
                # The constant strategy has a specific behavior and preserve empty
                # features even with ``keep_empty_features=False``. We need to drop
                # the column for consistency.
                # TODO: remove this `if` branch once the following issue is addressed:
                # https://github.com/scikit-learn/scikit-learn/issues/29827
                X_filled = X_filled[:, ~self._is_empty_feature]
        else:
            # mark empty features as not missing and keep the original imputation
            mask_missing_values[:, self._is_empty_feature] = False
            Xt = X
            Xt[:, self._is_empty_feature] = X_filled[:, self._is_empty_feature]
```

We don't really need to expose it so I think that we can keep it private by starting with an `_`.

arifqodari replied:
Addressed in https://github.com/scikit-learn/scikit-learn/pull/29779/commits/3d04df9fdb51b8c8f20e0cf1e45860470171c82b

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -1786,3 +1786,57 @@ def test_simple_imputer_constant_fill_value_casting():
+    X_imputed_false = imputer_false.fit_transform(X)
+
+    imputer_true = IterativeImputer(
+        initial_strategy=strategy, fill_value=1, keep_empty_features=True
```
I would still use `fill_value=0` to not have a special casing for `constant` when checking the value.

arifqodari replied:
Addressed in https://github.com/scikit-learn/scikit-learn/pull/29779/commits/31cf853d7cf4ed09daa1555cc1d5b3196ee1b681

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -1786,3 +1786,57 @@ def test_simple_imputer_constant_fill_value_casting():
+
+
+@pytest.mark.parametrize("strategy", ["mean", "median", "most_frequent", "constant"])
+def test_iterative_imputer_no_empty_features(strategy):
```
There is a test called `test_iterative_imputer_keep_empty_features`: we can remove it and make sure that the case that we checked there is also covered by one of the two cases here.

arifqodari replied:
Addressed in https://github.com/scikit-learn/scikit-learn/pull/29779/commits/0e1f067d1f7a51f8112e8ebb0d04d0f05916ad88

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -1786,3 +1786,57 @@ def test_simple_imputer_constant_fill_value_casting():
+    X = _sparse_random_matrix(10, 10, density=0.95, random_state=rng).toarray()
+    X[X == 0] = np.nan
+
+    imputer_false = IterativeImputer(
```
I think that we can do better naming instead of `imputer_false` and `imputer_true`: `imputer_keep_empty_features` and `imputer_drop_empty_features`.

arifqodari replied:
Addressed in https://github.com/scikit-learn/scikit-learn/pull/29779/commits/8a71a26d778593f383efa709ccc6ac569245df82

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -1786,3 +1786,57 @@ def test_simple_imputer_constant_fill_value_casting():
+    imputer_false = IterativeImputer(
+        initial_strategy=strategy, fill_value=1, keep_empty_features=False
+    )
+    X_imputed_false = imputer_false.fit_transform(X)
```
Here, we can avoid to have to find a variable name and directly check

```python
assert_allclose(
    imputer_drop_empty_features.fit_transform(X),
    imputer_keep_empty_features.fit_transform(X),
)
```

Do not use `assert_array_equal` with floating dtype.

arifqodari replied:
Addressed in https://github.com/scikit-learn/scikit-learn/pull/29779/commits/8a71a26d778593f383efa709ccc6ac569245df82

arifqodari committed:
Better Variable Naming in Tests

arifqodari committed:
Use Simple Handcrafted Matrix Instread of Generated Randomly

arifqodari commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -1786,3 +1786,57 @@ def test_simple_imputer_constant_fill_value_casting():
+    the same result regardless of whether `keep_empty_features`.
+    """
+    rng = np.random.RandomState(0)
+    X = _sparse_random_matrix(10, 10, density=0.95, random_state=rng).toarray()
```
Addressed in https://github.com/scikit-learn/scikit-learn/pull/29779/commits/af88e79c9f5987b3ce0c5b851d1bd5cdb7a45093

arifqodari committed:
Black Formatting

arifqodari commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -1786,3 +1786,57 @@ def test_simple_imputer_constant_fill_value_casting():
+    )
+    X_imputed_true = imputer_true.fit_transform(X)
+
+    if strategy == "constant":
```
We will have different shapes of `X` depending on `initial_strategy` and `keep_empty_features`, right?

For example using this data
```python
    X = np.array(
        [
            [np.nan, np.nan, 0, 1],
            [np.nan, 2, np.nan, 3],
            [np.nan, 4, 5, np.nan]
        ]
    )
```

**Both `X`s have the same shape when initial_strategy = constant**
- keep_empty_features == False. X_imputed.shape == (3, 4)
- keep_empty_features == True. X_imputed.shape == (3, 4)

In this case we just need one test
```python
assert_allclose(X_drop_empty_features, X_keep_empty_features, rtol=1e-4)
```

**But `X`s will have different shapes when initial_strategy != constant**
- keep_empty_features == False. X_imputed.shape == (3, 3)
- keep_empty_features == True. X_imputed.shape == (3, 4)

I guess in this case we still need to check whether strategy is constant or not, then continue to test equality on non-empty features. What do you think?

glemaitre replied:
Not if you take this fix: https://github.com/scikit-learn/scikit-learn/pull/29779#discussion_r1752191625

It will drop the feature even with the `constant` strategy.

glemaitre replied:
> I see. So, the expected condition is to drop the empty features, even with constant strategy.

Yes we should be consistent. In `SimpleImputer`, it was for backward compatibility but I think that we should deprecate and so I open an issue.

arifqodari replied:
Addressed in https://github.com/scikit-learn/scikit-learn/pull/29779/commits/3d04df9fdb51b8c8f20e0cf1e45860470171c82b

arifqodari committed:
Simplify Test Iterative Imputer with Empty Features

arifqodari committed:
Fix Missing Value Masking

arifqodari committed:
Remove Redundant Test

arifqodari committed:
Add Test Data along with Few More Tests

arifqodari commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -1786,3 +1786,57 @@ def test_simple_imputer_constant_fill_value_casting():
+    rng = np.random.RandomState(0)
+    X = _sparse_random_matrix(10, 10, density=0.95, random_state=rng).toarray()
+    X[X == 0] = np.nan
+    X[:, 0] = np.nan
```
Addressed in https://github.com/scikit-learn/scikit-learn/pull/29779/commits/a3bbbdafd268c72298b87de1564f3ef43b4eb43f

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -1786,3 +1768,57 @@ def test_simple_imputer_constant_fill_value_casting():
+
+@pytest.mark.parametrize("strategy", ["mean", "median", "most_frequent", "constant"])
+def test_iterative_imputer_no_empty_features(strategy):
+    """Check the behavior of the iterative imputer.
+    When the input matrix contains no empty features, it should return
+    the same result regardless of whether `keep_empty_features`.
+    """
```
```suggestion
    """Check the behaviour of `keep_empty_features` with no empty features.

    With no-empty features, we should get the same imputation whatever the
    parameter `keep_empty_features`.

    Non-regression test for:
    https://github.com/scikit-learn/scikit-learn/issues/29375
    """
```

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -1786,3 +1768,57 @@ def test_simple_imputer_constant_fill_value_casting():
+
+@pytest.mark.parametrize("strategy", ["mean", "median", "most_frequent", "constant"])
+def test_iterative_imputer_with_empty_features(strategy):
+    """Check the behavior of the iterative imputer.
+    When the input matrix contains empty features, it should return
+    the same values for the non-empty features regardless of
+    `keep_empty_features`. Additionally, the empty features
+    should become 0s when `keep_empty_features` is set to True.
+    """
```
```suggestion
    """Check the behaviour of `keep_empty_features` in the presence of empty features.

    With `keep_empty_features=True`, the empty feature will be imputed with the value
    defined by the intial imputation.

    Non-regression test for:
    https://github.com/scikit-learn/scikit-learn/issues/29375
    """
```

glemaitre reviewed (commented):
A couple of additional small changes. But otherwise it looks good.

arifqodari committed:
Update docstring for Test without Empty Feature

Co-authored-by: Guillaume Lemaitre <guillaume@probabl.ai>

arifqodari committed:
Update docstring for Test with Empty Feature

Co-authored-by: Guillaume Lemaitre <guillaume@probabl.ai>

arifqodari committed:
Parametrize on X_test for test_iterative_imputer_with_empty_features

arifqodari commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -1786,3 +1768,57 @@ def test_simple_imputer_constant_fill_value_casting():
+    X_train = np.array(
+        [[np.nan, np.nan, 0, 1], [np.nan, 2, np.nan, 3], [np.nan, 4, 5, np.nan]]
+    )
+    X_test = np.array([[1, 2, 3, 4], [5, 6, 7, 8]])
```
Addressed in https://github.com/scikit-learn/scikit-learn/pull/29779/commits/004df0c5b983551ab7f209cbee35f95bdeef9df9

arifqodari committed:
Merge branch 'main' into fix_iterative_imputer_keep_empty_features

glemaitre commented on sklearn/impute/tests/test_impute.py:
```diff
@@ -1786,3 +1768,70 @@ def test_simple_imputer_constant_fill_value_casting():
+    """Check the behaviour of `keep_empty_features` in the presence of empty features.
+
+    With `keep_empty_features=True`, the empty feature will be imputed with the value
+    defined by the intial imputation.
```
```suggestion
    defined by the initial imputation.
```

arifqodari committed:
Fix Typo in Docstring

Co-authored-by: Guillaume Lemaitre <guillaume@probabl.ai>

jeremiedbb reviewed (approved):
LGTM. Thanks @arifqodari

jeremiedbb commented:
Let's merge and hopefully we won't forget to clean it up when we fix the inconsistency in the SimpleImputer directly :smile:
